File: src/pub_node_cpp/src/publisher.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/float32.hpp"
#include <chrono>

using namespace std::chrono_literals;

class PublisherObject : public rclcpp::Node
{
public:
    PublisherObject() : Node("cpp_publisher")
    {
        steer_pub_ = this->create_publisher<std_msgs::msg::Float32>("steer", 10);
        throttle_pub_ = this->create_publisher<std_msgs::msg::Float32>("throttle", 10);
        timer_ = this->create_wall_timer(1000ms, std::bind(&PublisherObject::timer_callback, this));
    }

private:
    void timer_callback()
    {
        std_msgs::msg::Float32 steer_msg;
        std_msgs::msg::Float32 throttle_msg;

        steer_msg.data = 0.15;
        throttle_msg.data = -0.3;

        RCLCPP_INFO(this->get_logger(), "Publishing steer: %.2f, throttle: %.2f", steer_msg.data, throttle_msg.data);

        steer_pub_->publish(steer_msg);
        throttle_pub_->publish(throttle_msg);
    }

    rclcpp::Publisher<std_msgs::msg::Float32>::SharedPtr steer_pub_;
    rclcpp::Publisher<std_msgs::msg::Float32>::SharedPtr throttle_pub_;
    rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<PublisherObject>());
    rclcpp::shutdown();
    return 0;
}

File: src/joystick_control/joystick_control/pub_node/publisher.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import Float32
from piracer.gamepads import ShanWanGamepad

class PublisherObject(Node):
    def __init__(self):
        super().__init__('publisher')

        self.gamepad = ShanWanGamepad()

        self.steer_msg = self.create_publisher(Float32, 'steer', 10)
        self.throttle_msg = self.create_publisher(Float32, 'throttle', 10)
        self.timer = self.create_timer(0.2, self.timer_callback)
    
    
    def timer_callback(self):
        gamepad_input = self.gamepad.read_data()

        throttle_msg = Float32()
        steer_msg = Float32()    

        throttle_msg.data = gamepad_input.analog_stick_right.y
        steer_msg.data = gamepad_input.analog_stick_left.x

        self.steer_msg.publish(steer_msg)
        self.throttle_msg.publish(throttle_msg)


def main(args=None):
    rclpy.init(args=args)
    node = PublisherObject()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()